benches/hash.rs

```diff
@@ -1,116 +1,110 @@
+    //
+    // Also note that the single entry case does not have a loop, so this is not directly
+    // comparable to the iterations bellow.
+    let s: [RpoDigest; 2] = [Rpo256::hash(&[1_u8]), Rpo256::hash(&[2_u8])];
+    group.bench_with_input(BenchmarkId::new("RPO256", 1), &s, |b, i| {
+        b.iter(|| {
+            Rpo256::merge(black_box(&i));
+        })
     });
 
-    c.bench_function("RPO256 sequential hashing (random)", |bench| {
-        bench.iter_batched(
-            || {
-                let v: [Felt; 100] = (0..100)
-                    .into_iter()
-                    .map(|_| Felt::new(rand_value()))
-                    .collect::<Vec<Felt>>()
-                    .try_into()
-                    .expect("should not fail");
-                v
-            },
-            |state| Rpo256::hash_elements(&state),
-            BatchSize::SmallInput,
-        )
-    });
-}
-
-fn blake3_2to1(c: &mut Criterion) {
-    let v: [<Blake3_256 as Hasher>::Digest; 2] =
+    let s2: [<Blake3_256 as Hasher>::Digest; 2] =
         [Blake3_256::hash(&[1_u8]), Blake3_256::hash(&[2_u8])];
-    c.bench_function("Blake3 2-to-1 hashing (cached)", |bench| {
-        bench.iter(|| Blake3_256::merge(black_box(&v)))
+    group.bench_with_input(BenchmarkId::new("Blake3/256", 1), &s2, |b, i| {
+        b.iter(|| {
+            Blake3_256::merge(black_box(&i));
+        })
     });
 
-    c.bench_function("Blake3 2-to-1 hashing (random)", |bench| {
-        bench.iter_batched(
-            || {
-                [
-                    Blake3_256::hash(&rand_value::<u64>().to_le_bytes()),
-                    Blake3_256::hash(&rand_value::<u64>().to_le_bytes()),
-                ]
+    // Benchmark `merge` in a hot loop with varying number of elements, the goal is to spend more
+    // time in the hashing function so that the cost of measuring it will dissipiate in the
+    // background and give better view of the algorithms performance.
+    for size in BATCH_SIZES {
+        let v: Vec<[RpoDigest; 2]> = (0..size)
+            .map(|_| {
+                let left = rand_array::<Felt, 4>().into();
+                let right = rand_array::<Felt, 4>().into();
+                [left, right]
+            })
+            .collect();
+
+        group.bench_with_input(
+            BenchmarkId::new(format!("RPO256/{size}"), size),
+            &v,
+            |b, i| {
+                b.iter(|| {
+                    for pair in i {
+                        Rpo256::merge(black_box(&pair));
+                    }
+                })
             },
-            |state| Blake3_256::merge(&state),
-            BatchSize::SmallInput,
-        )
-    });
+        );
+
+        let v2: Vec<[<Blake3_256 as Hasher>::Digest; 2]> = v
+            .iter()
+            .map(|[l, r]| [l.as_bytes().into(), r.as_bytes().into()])
+            .collect();
+        group.bench_with_input(
+            BenchmarkId::new(format!("Blake3/256/{size}"), size),
```

Comment:
I think doing `/{size}` is redundant here. We could probably just do: `BenchmarkId::new("Blake3_256", size)`.